confluent_kafka/producer.py

```diff
@@ -0,0 +1,130 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from .cimpl import Producer as _impl, PARTITION_UA
+from warnings import warn
+
+
+class Producer(_impl):
+    """
+        Create a new Kafka Producer instance.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+            calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties. At a minimum ``bootstrap.servers`` **should** be set.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_serializer(topic, key): Converts key to bytes.
+            **note** serializers are responsible for handling NULL keys
+        :param func value_serializer(topic, value): Converts value to bytes.
+            **note** serializers are responsible for handling NULL keys
+        :param func error_cb(kafka.KafkaError): Callback for generic/global error events.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict object.
+    """
+
+    __slots__ = ["_key_serializer", "_value_serializer"]
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_serializer=None, value_serializer=None,
+                 error_cb=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_serializer = key_serializer
+        self._value_serializer = value_serializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
```

Comment:
same comment as before (why allow them in config?

Reply:
backward compatibility